lib/madwf_ml.cpp

```diff
@@ -0,0 +1,347 @@
+    param(solve_param.madwf_param),
+    mu(param.madwf_diagonal_suppressor),
+    prec_precondition(solve_param.precision_precondition),
+    profile(profile)
+  {
+    if (getVerbosity() >= QUDA_VERBOSE) {
+      printfQuda("Launching MADWF accelerator ... \n");
+      printfQuda("madwf_mu (low modes suppressor)                   = %.4f\n", param.madwf_diagonal_suppressor);
+      printfQuda("madwf_ls (cheap Ls)                               = %d\n", param.madwf_ls);
+      printfQuda("madwf_null_miniter                                = %d\n", param.madwf_null_miniter);
+      printfQuda("madwf_null_tol                                    = %4.2e\n", param.madwf_null_tol);
+      printfQuda("madwf_train_maxiter (max # of iters for training) = %d\n", param.madwf_train_maxiter);
+    }
+  }
+
+  void MadwfAcc::fill_random(std::vector<transfer_float> &v)
+  {
+    static std::random_device rd;
+    // the good rng
+    static std::mt19937 rng(23ul * comm_rank());
+    // The gaussian distribution
+    static std::normal_distribution<double> n(0., 1.);
+
+    for (auto &x : v) { x = 1e-1 * n(rng); }
+  }
+
+  void MadwfAcc::apply(Solver &base, ColorSpinorField &out, const ColorSpinorField &in)
+  {
+    madwf_ml::transfer_5d_hh(*forward_tmp, in, device_param, false);
+    pushVerbosity(QUDA_SILENT);
+    base(*backward_tmp, *forward_tmp);
+    popVerbosity();
+    madwf_ml::transfer_5d_hh(out, *backward_tmp, device_param, true);
+
+    blas::axpy(mu, const_cast<ColorSpinorField &>(in), out);
+  }
+
+  double MadwfAcc::cost(const DiracMatrix &ref, Solver &base, ColorSpinorField &out, const ColorSpinorField &in)
+  {
+
+    ColorSpinorParam csParam(in);
+    ColorSpinorField tmp1(csParam);
+    ColorSpinorField tmp2(csParam);
+
+    apply(base, tmp1, in);
+    ref(tmp2, tmp1);
+
+    blas::copy(out, in);
+
+    // M * T^ * A * T * phi - phi
+    return blas::xmyNorm(tmp2, out);
+  }
+
+  void MadwfAcc::train(const DiracMatrix &ref, Solver &base, Solver &null, const ColorSpinorField &in,
+                       bool tune_suppressor)
+  {
+
+    constexpr int complex_matrix_size = static_cast<int>(transfer_t); // spin by spin
+
+    int Ls = in.X(4);
+    int Ls_base = param.madwf_ls;
+    int param_size = Ls * Ls_base * complex_matrix_size * 2;
+    std::vector<transfer_float> host_param(param_size);
+
+    if (param.madwf_param_load) {
+      load_parameter(Ls, Ls_base, param_size);
+
+      ColorSpinorParam csParam(in);
+      csParam.x[4] = Ls_base;
+      csParam.create = QUDA_NULL_FIELD_CREATE;
+      csParam.setPrecision(prec_precondition);
+
+      forward_tmp = std::make_unique<ColorSpinorField>(csParam);
+      backward_tmp = std::make_unique<ColorSpinorField>(csParam);
+
+      return;
+    }
+
+    profile.TPSTART(QUDA_PROFILE_INIT);
+
+    ColorSpinorParam csParam(in);
+    ColorSpinorField null_x(csParam);
+    ColorSpinorField null_b(csParam);
+
+    RNG rng(null_b, 1234);
```

Comment:
Maybe orthogonal to this PR, but we use `1234` as the seed in _so many_ places in QUDA. I think we should at least use a different number here, and in the future make the seed something we specify externally and pass in.

Reply:
https://github.com/lattice/quda/pull/1061/commits/65dc02461126ad5aee630ce9153055476d15d98b